### PR TITLE
feat(metrics): loading strategy timing/count

### PR DESCRIPTION
**Pull request by yixinh-nv**
Followup on the PR 706. This PR implements the load strategy chain phase. 

| Family | Labels | Answers |
| --- | --- | --- |
| `mx_load_strategy_seconds` | `engine`, `model`, `strategy`, `outcome`, `scheme` | How long the strategy spent in the chain phase? |
| `mx_load_strategy_skipped_total` | `engine`, `strategy`, `scheme` | Which never got to try? |

## Benchmarking Results

Three pods on H200s, one GPU each: two model sizes under SGLang, plus a vLLM
pod loading the same 0.5B as one of them.

| Engine | Model | Strategy | Outcome | Mean |
| --- | --- | --- | --- | ---: |
| sglang | `Qwen2.5-0.5B-Instruct` | `default` | success | 9.245 s |
| sglang | `Qwen2.5-7B-Instruct` | `default` | success | 38.405 s |
| vllm | `Qwen2.5-0.5B-Instruct` | `instant_tensor` | fallback_dirty | 36.50 ms |
| vllm | `Qwen2.5-0.5B-Instruct` | `default` | success | 4.548 s |

The nesting holds on every pod — `sum(strategy) <= chain <= load`:

| Engine | Model | Σ strategies | `chain` | load | Σ / `chain` |
| --- | --- | ---: | ---: | ---: | ---: |
| sglang | `Qwen2.5-0.5B-Instruct` | 9.245 s | 9.248 s | 9.252 s | 99.96% |
| sglang | `Qwen2.5-7B-Instruct` | 38.405 s | 38.409 s | 38.412 s | 99.99% |
| vllm | `Qwen2.5-0.5B-Instruct` | 4.585 s | 4.589 s | 5.609 s | 99.91% |

SGLang's `transfer_engine` transport never enters the chain, so it recorded
`mx_load_seconds` and nothing beneath it. It is a chain by hand — try the
transfer, on failure re-initialize and load natively — so it now records the
same `chain` and `publish` phases and the same attempt spans (`transfer_engine`,
then `default` when it misses) from the same context managers. 28 lines added
in the loader, all `with` blocks; no interface touched. Not exercised on
hardware: it needs a mooncake peer, so the screenshot has no row for it.

model_load

## Summary by CodeRabbit

- **New Features**
  - Added Prometheus metrics for model-load strategy duration, outcomes, and skipped-strategy reasons.
  - Added engine labels for SGLang, TRT-LLM, and vLLM load metrics.
  - Improved load-strategy reporting with detailed outcomes, including fallback and recovery results.
  - Added dashboard panels for strategy attempts, durations, outcomes, and skip reasons.

- **Documentation**
  - Expanded metrics documentation with interpretation guidance, invariants, and dashboard updates.
  - Documented the required process for keeping metrics, dashboards, and alert rules synchronized.

### CODE REVIEW

File: docs/METRICS.md [hunk 1]
```diff
@@ -12,13 +12,13 @@ deployment.
 
 This documents what ships today: the exposition path, `mx_build_info`, per-RPC
 and storage-backend coverage on the server, the download lifecycle, NIXL client
-health, the Kubernetes scrape, alerting and dashboard surface, and the first two
-load-timing tiers — the `load_model()` window and the phases that partition it.
+health, the Kubernetes scrape, alerting and dashboard surface, and the first
+three load-timing tiers — the `load_model()` window, the phases that partition
+it, and the strategy attempts that partition the `chain` phase.
```
[the remark this answers is not in the document]
Comment (yixinh-nv) [reply docs/METRICS.md]: Fixed in 4435bba7 — both summary lines now say the attempts run inside the `chain` phase, which is what the detailed section already said.

File: modelexpress_client/python/modelexpress/load_strategy/__init__.py [hunk 2]
```diff
@@ -88,18 +99,34 @@ def run(model: nn.Module, ctx: LoadContext) -> nn.Module:
...
                     return result.value
                 except StrategyRecoveryError:
                     # Recovery already failed, so no later strategy can safely
                     # use the current model. Fail closed and retain the original
                     # recovery error as the exception cause.
+                    outcome = "recovery_error"
                     strategy.rollback(ctx)
                     raise
                 except StrategyFailed as e:
+                    outcome = "fallback_dirty" if e.mutated else "fallback"
```
[the remark this answers is not in the document]
Comment (yixinh-nv) [reply modelexpress_client/python/modelexpress/load_strategy/__init__.py]: Correct and fixed in 4435bba7 — `recovery_error` is now assigned before rollback and re-init and downgraded only once they complete, with parametrized coverage for both failure paths that fails against the previous ordering.

File: modelexpress_client/python/modelexpress/load_strategy/model_streamer_strategy.py [hunk 3]
```diff
@@ -49,22 +49,23 @@ class ModelStreamerStrategy(LoadStrategy):
...
+            return "package_missing"
 
         model_uri = envs.MX_MODEL_URI or ""
         if not model_uri:
             logger.info(
                 f"[Worker {ctx.global_rank}] MX_MODEL_URI not set, skipping model streamer"
             )
-            return False
-        return True
+            return "not_configured"
```
[the remark this answers is not in the document]
Comment (yixinh-nv) [reply modelexpress_client/python/modelexpress/load_strategy/model_streamer_strategy.py]: Real bug, but pre-existing and not from this PR — `main` has the same `envs.MX_MODEL_URI` gate against a `load()` that resolves from `ctx.model_config`, and this PR only changed the return value from `False` to `"not_configured"`; since fixing it makes the strategy newly eligible in deployments where it is skipped today, that behavior change belongs in its own PR rather than a metrics one.

File: modelexpress_client/python/modelexpress/load_strategy/base.py [hunk 4]
```diff
@@ -75,14 +75,35 @@ class LoadStrategy(ABC):
     name: str
     requires: ClassVar[tuple] = ()
 
-    def is_available(self, ctx: LoadContext) -> bool:
-        """Check environment: is this strategy usable right now?"""
+    def skip_reason(self, ctx: LoadContext) -> str | None:
```
Comment (zhengluo-nv) [inline modelexpress_client/python/modelexpress/load_strategy/base.py]: This is too intrusive to the existing core logic. We should not update the core interface just for adding a metric. Please try to modularize your metric publisher logic and minimize the integration surface

Comment (yixinh-nv) [reply modelexpress_client/python/modelexpress/load_strategy/base.py]: Agreed and reverted in 71735741 — `base.py` and all six strategies are back to the `is_available` contract on `main` (along with the ~37 test patch targets that had migrated with it), and L2 now sits behind two calls into `metrics.py`: `record_chain_skips` and a `time_load_strategy` context manager, with the counter losing its `reason` label since carrying that was the only thing the interface change bought.
